conans/client/rest/rest_client_v1.py

```diff
@@ -251,19 +251,20 @@ def upload_package(self, package_reference, the_files, retry, retry_wait, no_ove
         t1 = time.time()
         # Get the remote snapshot
         remote_snapshot = self._get_package_snapshot(package_reference)
-        local_snapshot = {filename: md5sum(abs_path) for filename, abs_path in the_files.items()}
+        if remote_snapshot:
+            remote_manifest = self.get_package_manifest(package_reference)
+            local_manifest = FileTreeManifest.loads(load(the_files["conanmanifest.txt"]))
 
-        # Get the diff
-        new, modified, deleted = diff_snapshots(local_snapshot, remote_snapshot)
-        if not new and not deleted and modified in (["conanmanifest.txt"], []):
-            return False
+            if remote_manifest == local_manifest:
+                return False
 
-        if no_overwrite and remote_snapshot:
             if no_overwrite == "all":
                 raise ConanException("Local package is different from the remote package. "
                                      "Forbidden overwrite")
 
-        files_to_upload = {filename: the_files[filename] for filename in new + modified}
+        local_snapshot = {filename: None for filename, abs_path in the_files.items()}
```

Comment:
`local_snapshot = set(the_files.keys())` (or leave it as an iterable, same as above)